pycrdt_websocket/awareness.py

```diff
@@ -56,10 +97,70 @@ def get_changes(self, message: bytes) -> dict[str, Any]:
                     if state != prev_state:
                         filtered_updated.append(client_id)
                     updated.append(client_id)
-        return {
+
+        changes = {
             "added": added,
             "updated": updated,
             "filtered_updated": filtered_updated,
             "removed": removed,
             "states": states,
         }
+
+        # Do not trigger the callbacks if it is only a keep alive update
+        if added or filtered_updated or removed:
+            for callback in self._subscriptions:
+                callback(changes)
+
+        return changes
+
+    def get_local_state(self) -> dict[str, Any]:
+        return self._states.get(self.client_id, {})
+
+    def set_local_state(self, state: dict[str, Any]) -> None:
+        # Update the state and the meta.
+        timestamp = int(time.time() * 1000)
+        clock = self.meta.get(self.client_id, {}).get("clock", -1) + 1
+        self._states[self.client_id] = state
+        self.meta[self.client_id] = {"clock": clock, "last_updated": timestamp}
+        # Build the message to broadcast, with the following information:
+        # - message type
+        # - length in bytes of the updates
+        # - number of updates
+        # - for each update
+        #   - client_id
+        #   - clock
+        #   - length in bytes of the update
+        #   - encoded update
+        msg = json.dumps(state, separators=(",", ":")).encode("utf-8")
+        msg = write_var_uint(len(msg)) + msg
+        msg = write_var_uint(clock) + msg
+        msg = write_var_uint(self.client_id) + msg
+        msg = write_var_uint(1) + msg
+        msg = write_var_uint(len(msg)) + msg
+        msg = write_var_uint(YMessageType.AWARENESS) + msg
+
+        if self.on_change:
+            with start_blocking_portal() as portal:
```

Comment:
But all this happens in the same thread, right? `start_blocking_portal()` is for spawning tasks from a worker thread.